File: 0309_Best_Time_to_Buy_and_Sell_Stock_with_Cooldown/code4.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        
        n = len(prices)
        
        maxScore = {}
        def getMaxScore(index):
            
            if index not in maxScore:
                
                if index >= n - 1:
                    return 0

                maxScore[index] = max(getMaxScore(index + 1), prices[index + 1] + recur(index + 3))
                
            return maxScore[index]
        
        memo = {}
        def recur(index):
            
            if index not in memo:
                
                if index >= n - 1:
                    return 0
                
                memo[index] = max(recur(index + 1), getMaxScore(index) - prices[index])
                
            return memo[index]
        
        return recur(0)

File: 0714_Best_Time_to_Buy_and_Sell_Stock_with_Transaction_Fee/code4.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) {
        int n = prices.size();
        vector<vector<int>> memo(n, vector<int>(2, -1));
        return recur(0, false, memo, prices, fee);
    }
    
    int recur(int idx, bool isHold, vector<vector<int>>& memo, vector<int>& prices, int fee){
        int n = prices.size();
        
        if(idx >= n){
            return 0;
        }
        
        if(memo[idx][isHold] != -1){
            return memo[idx][isHold];
        }
        
        int ans = INT_MIN;
        if(isHold){
            ans = max(prices[idx] - fee + recur(idx + 1, false, memo, prices, fee), recur(idx + 1, true, memo, prices, fee));
        } else {
            ans = max(-prices[idx] + recur(idx + 1, true, memo, prices, fee), recur(idx + 1, false, memo, prices, fee));
        }
        
        memo[idx][isHold] = ans;
        return ans;
    }
    
};